File: MergeSort.cpp
#include<stdio.h>

   /**
  　　* 0 <= p <= q < r, subarray array[p..q] and array[q+1..r] are already sorted.
  　　* the merge() function merges the two sub-arrays into one sorted array.
    */

   __int64 count;
   __int64 a[500005];
   __int64 temp[500005];
   //防止数据溢出

void Merge(int p, int q,int r)
{
        int i,k;
        int begin1,end1,begin2,end2;
        k=p;
        begin1 = p;
        end1 = q;
        begin2 = q+1;
        end2 = r;
        while((begin1 <=end1)&&( begin2<=end2))
            {
                if(a[begin1]<=a[begin2])
                {
                    temp[k] = a[begin1];
                    begin1++;
                }
                else
                {
                    temp[k] = a[begin2];
                    begin2++;

              //count++;
              /*
              *
              *求逆序对，当array[begin1]中的数比array[begin2]中的数大时；
             *那么可以推出此数后面的数和array[begin2]这个数大;
              */
                    count += end1 - begin1+1;

                }

                k++;

            }

            while(begin1<=end1)
            {
                temp[k++]=a[begin1++];

            }

            while(begin2<=end2)
            {
                temp[k++]=a[begin2++];

            }

            for(i=p;i<=r;i++)       //把合并后的数组还原

            a[i]=temp[i];

}



  /**
  归并操作（二分法）
  */

void MergeSort(int first,int last)
{
    int mid=0;
    if(first<last)
    {
        mid=(first+last)/2;
        MergeSort(first,mid);
        MergeSort(mid+1,last);
        Merge(first,mid,last);
    }

}


int main()
{
    int i;
    int input;
    scanf("%d",&input);
    while(input!=0)
    {
        count=0;
        for(i=0;i<input;i++)
    {
        scanf("%I64d",&a[i]);
    }

    MergeSort(0,input-1);

    printf("%I64d\n",count);

    scanf("%I64d",&input);
    }

 return 1;
 }

File: MergeSORT.java
package mergeSort;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.StreamTokenizer;

public class MergeSORT {
	static long count = 0;
    public static void mergeSort(int []input,int left,int right){
        //分治
    	int mid = (left + right) / 2;  
        if (left < right) {  
            // 左边  
        	
            mergeSort(input,left,mid);  
            // 右边  
            
            mergeSort(input,mid+1,right);  
            // 左右归并  
            
            merge(input,left,mid,right);  
            }  
        }
    /**
     * 二路归并
     * input
     * left左数组的第一个元素的索引
     * center左数组的最后一个元素的索引，center+1是右数组第一个元素的索引
     * right右数组最后一个元素的索引
     */
    public static void merge(int[]input,int left,int center,int right){
    	int []tempArray = new int[right-left+1];
        int mid = center+1;
        int temp = left;
        int current = 0;
        while(left<=center && mid<=right){
            if(input[left]>input[mid]){
                tempArray[current++]=input[mid++];
                /*
                 * 如果input[left]>input[mid]，那么表明下标为left~center的值都会比input[mid]大;
                 * (5,7,8)(4,9)合并为 3+0+0+0
                 * */               
                count+=center-left+1;
            }else{
                tempArray[current++]=input[left++];
            }
        }
        //只会执行一个
        while(left<=center){
            tempArray[current++]=input[left++];
        }
        while(mid<=right){
            tempArray[current++]=input[mid++];
        }
        current=0;
        while(temp<=right){
            input[temp++]=tempArray[current++];
        }
    }
    public static void main(String[] args) throws IOException {
        StreamTokenizer cin = new StreamTokenizer(new BufferedReader(new InputStreamReader(System.in)));
        while(cin.nextToken()!=cin.TT_EOF){
            int n = (int)cin.nval;
            int []input = new int[n];
            for(int i=0;i<n;i++){
                cin.nextToken();
                input[i] = (int)cin.nval;
            }
            count=0;
            mergeSort(input,0,input.length-1);
            System.out.println(count);
        }
    }
}
